From 7bcf5f656dcbea872ab5b854a01dd3c0801f1298 Mon Sep 17 00:00:00 2001
From: Gajanan Bhat <gajanan.bhat@intel.com>
Date: Mon, 12 Oct 2015 12:28:09 +0530
Subject: [PATCH] REVERTME [VPG]: drm/i915: Disable maxfifo if zorder needs to
 be updated

In Cherryview if display is in maxfifo mode, zorder update does not take effect.
We need to exit from maxfifo mode if enabled before updating zoder. This patch
checks if in set display function zorder update was requested and disables
maxfifo if needed.

REVERTME: CHT product kernel relies on internal implementation of
atomic flip. Atomic flip is already addressed in a clean fashion using
nuclear flip in upstream and is used for Broxton and beyond.
So this patch will be a REVERTME patch and can be discarded in gmin_dev &
broxton trees and future forklifts.

Change-Id: I1eccd6dfefd9aea4a703785aac36a2fb190b63ee
Tracked-On: https://jira01.devtools.intel.com/browse/OAM-5799
Signed-off-by: Gajanan Bhat <gajanan.bhat@intel.com>
---
 drivers/gpu/drm/i915/intel_display.c | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/drivers/gpu/drm/i915/intel_display.c b/drivers/gpu/drm/i915/intel_display.c
index 9adec40..e3d5c5c 100644
--- a/drivers/gpu/drm/i915/intel_display.c
+++ b/drivers/gpu/drm/i915/intel_display.c
@@ -11394,10 +11394,15 @@ static int intel_crtc_set_display(struct drm_crtc *crtc,
 			plane_cnt++;
 	}
 
-	/* Disable maxfifo when moving from single plane to multiple plane */
+	/*
+	 * Disable maxfifo when moving from single plane to multiple plane or
+	 * when zorder is updated. Zorder cannot be updated in maxfifo mode.
+	 */
 	if (dev_priv->maxfifo_enabled && ((plane_cnt > 1) ||
 				!single_pipe_enabled(pipe_stat) ||
-				(pipe_stat & PIPE_ENABLE(PIPE_C)))) {
+				(pipe_stat & PIPE_ENABLE(PIPE_C) ||
+				(disp->update_flag &
+					DRM_MODE_SET_DISPLAY_UPDATE_ZORDER)))) {
 		intel_update_maxfifo(dev_priv, crtc, false);
 		dev_priv->wait_vbl = true;
 		dev_priv->vblcount =
-- 
1.9.1

